### PR TITLE
Support for IAM in pubsub

### LINKED ISSUE #680: Snippet generation for list methods includes cancellation token overloads
We don't include a cancellation token overload in our generated *actual* code for list methods, but the generated snippet refers to this phantom method. Examples of errors from snippet consumption:

```text
PublisherClientSnippets.g.cs:125: Member ID 'ListTopicsAsync(string,string,int?,CancellationToken)' matches no members.
PublisherClientSnippets.g.cs:191: Member ID 'ListTopicSubscriptionsAsync(string,string,int?,CancellationToken)' matches no members.
SubscriberClientSnippets.g.cs:92: Member ID 'ListSubscriptionsAsync(string,string,int?,CancellationToken)' matches no members.
```

### PR DESCRIPTION
**Pull request by chrisdunelm**
And fixes #680.
And corrects some missing imports.
And prefixes the snippet class/file name with 'Generated'.

### CODE REVIEW

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicClientTransformer.java [hunk 1]
```diff
@@ -218,6 +222,26 @@ private StaticLangXSettingsView generateXSettings(SurfaceTransformerContext cont
     return settingsMembers;
   }
 
+  private List<ReroutedGrpcView> generateReroutedGrpcView(SurfaceTransformerContext context) {
+    SurfaceNamer namer = context.getNamer();
+    ModelTypeTable typeTable = context.getTypeTable();
+    Set<ReroutedGrpcView> rerouteds = new HashSet<>();
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicClientTransformer.java]: reroutedViews?

Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicClientTransformer.java]: And perhaps use a LinkedHashSet so the iteration order is predictable? We wouldn't want generating the code twice to give different results. (I'm guessing the order is used later - I haven't checked yet.)

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpModelTypeNameConverter.java [hunk 2]
```diff
@@ -127,6 +130,14 @@ public TypeName getTypeName(ProtoElement elem) {
       prefix = parentEl.getSimpleName() + ".Types." + prefix;
       parentEl = parentEl.getParent();
     }
+    if (parentEl instanceof ProtoFile) {
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpModelTypeNameConverter.java]: Is this part of the IAM stuff, or fixing imports?

File: src/test/java/com/google/api/codegen/testdata/csharp_gapic_snippets_library.baseline [hunk 3]
```diff
@@ -1,4 +1,4 @@
-============== file: Google/Example/Library/V1/LibraryServiceClientSnippets.g.cs ==============
+============== file: Google/Example/Library/V1/GeneratedLibraryServiceClientSnippets.g.cs ==============
```
Comment (jskeet) [inline src/test/java/com/google/api/codegen/testdata/csharp_gapic_snippets_library.baseline]: Can we keep the filename as LibraryServiceClientSnippets.g.cs, but change the class name?
That's the way I've got it now - it's a little odd, but it gives prominence to the manual code, while still making the generated code available easily. Not a problem if it's hard though.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicClientTransformer.java (hunk 1 above)
Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicClientTransformer.java]: Good point, changed to LinkedHashSet. And renamed the local.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpModelTypeNameConverter.java (hunk 2 above)
Comment (chrisdunelm) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpModelTypeNameConverter.java]: Fixing imports. I've added a line to the PR description.

File: src/test/java/com/google/api/codegen/testdata/csharp_gapic_snippets_library.baseline (hunk 3 above)
Comment (chrisdunelm) [inline src/test/java/com/google/api/codegen/testdata/csharp_gapic_snippets_library.baseline]: Done

**Comment by chrisdunelm** [comment]
PTAL

**Comment by chrisdunelm** [comment]
Really must learn to run tests before committing...

**Pull request merged**